File: core/internal/group/group.go
package group

import "errors"

var (
	ErrTooFewMembers          = errors.New("group requires at least 2 members")
	ErrNotMember              = errors.New("not a member of the group")
	ErrNotOwner               = errors.New("not an owner of the group")
	ErrCannotDemoteOtherOwner = errors.New("owner cannot demote another owner")
	ErrTargetNotMember        = errors.New("target is not a member")
	ErrNoOwner                = errors.New("cannot add member when group has no owners")
	ErrAlreadyMember          = errors.New("already a member of the group")
)

// Service applies domain rules (member count, leave, owner permissions, auto-promote) using Store.
type Service struct {
	store Store
}

// NewService returns a domain service that uses the given store.
func NewService(store Store) *Service {
	return &Service{store: store}
}

// CreateGroup creates a group with at least 2 members. Owners can be 0, 1, or all (2-person: 隠匿してよい).
func (s *Service) CreateGroup(members []string, ownerDIDs []string) (groupID string, err error) {
	if len(members) < 2 {
		return "", ErrTooFewMembers
	}
	return s.store.CreateGroup(members, ownerDIDs)
}

// Leave removes the member from the group. If the group becomes 1 member, it is dissolved (deleted).
func (s *Service) Leave(groupID string, memberDID string) error {
	g, err := s.store.GetGroup(groupID)
	if err != nil {
		return err
	}
	if g == nil {
		return ErrGroupNotFound
	}
	if !contains(g.Members, memberDID) {
		return ErrNotMember
	}
	newMembers := without(g.Members, memberDID)
	newOwners := without(g.Owners, memberDID)
	if len(newMembers) < 2 {
		return s.store.DeleteGroup(groupID)
	}
	if len(newOwners) == 0 {
		newOwners = autoPromoteOneExcluding(newMembers, "")
	}
	return s.store.UpdateGroup(groupID, newMembers, newOwners)
}

// Kick removes targetDID from the group. Caller must be an owner.
func (s *Service) Kick(groupID string, actorDID string, targetDID string) error {
	g, err := s.store.GetGroup(groupID)
	if err != nil {
		return err
	}
	if g == nil {
		return ErrGroupNotFound
	}
	if !contains(g.Owners, actorDID) {
		return ErrNotOwner
	}
	if !contains(g.Members, targetDID) {
		return ErrTargetNotMember
	}
	newMembers := without(g.Members, targetDID)
	newOwners := without(g.Owners, targetDID)
	if len(newMembers) < 2 {
		return s.store.DeleteGroup(groupID)
	}
	if len(newOwners) == 0 {
		newOwners = autoPromoteOneExcluding(newMembers, "")
	}
	return s.store.UpdateGroup(groupID, newMembers, newOwners)
}

// AppointOwner adds targetDID to owners. Caller must be an owner; target must be a member.
func (s *Service) AppointOwner(groupID string, actorDID string, targetDID string) error {
	g, err := s.store.GetGroup(groupID)
	if err != nil {
		return err
	}
	if g == nil {
		return ErrGroupNotFound
	}
	if !contains(g.Owners, actorDID) {
		return ErrNotOwner
	}
	if !contains(g.Members, targetDID) {
		return ErrTargetNotMember
	}
	if contains(g.Owners, targetDID) {
		return nil
	}
	newOwners := append([]string{}, g.Owners...)
	newOwners = append(newOwners, targetDID)
	return s.store.UpdateGroup(groupID, g.Members, newOwners)
}

// SelfDemote removes actorDID from owners. If that was the last owner, one remaining member is auto-promoted.
func (s *Service) SelfDemote(groupID string, actorDID string) error {
	g, err := s.store.GetGroup(groupID)
	if err != nil {
		return err
	}
	if g == nil {
		return ErrGroupNotFound
	}
	if !contains(g.Owners, actorDID) {
		return ErrNotOwner
	}
	newOwners := without(g.Owners, actorDID)
	if len(newOwners) == 0 {
		newOwners = autoPromoteOneExcluding(g.Members, actorDID)
	}
	return s.store.UpdateGroup(groupID, g.Members, newOwners)
}

// DemoteOwner removes targetDID from owners. Caller must be an owner. Fails if target is another owner (他オーナー降格不可).
func (s *Service) DemoteOwner(groupID string, actorDID string, targetDID string) error {
	g, err := s.store.GetGroup(groupID)
	if err != nil {
		return err
	}
	if g == nil {
		return ErrGroupNotFound
	}
	if !contains(g.Owners, actorDID) {
		return ErrNotOwner
	}
	if actorDID == targetDID {
		return s.SelfDemote(groupID, actorDID)
	}
	if contains(g.Owners, targetDID) {
		return ErrCannotDemoteOtherOwner
	}
	return nil
}

// AddMember adds newMemberDID to the group. Fails if the group has no owners (オーナー0では3人目招待不可).
func (s *Service) AddMember(groupID string, newMemberDID string) error {
	g, err := s.store.GetGroup(groupID)
	if err != nil {
		return err
	}
	if g == nil {
		return ErrGroupNotFound
	}
	if len(g.Owners) == 0 {
		return ErrNoOwner
	}
	if contains(g.Members, newMemberDID) {
		return ErrAlreadyMember
	}
	newMembers := append([]string{}, g.Members...)
	newMembers = append(newMembers, newMemberDID)
	return s.store.UpdateGroup(groupID, newMembers, g.Owners)
}

func contains(slice []string, x string) bool {
	for _, v := range slice {
		if v == x {
			return true
		}
	}
	return false
}

func without(slice []string, x string) []string {
	var out []string
	for _, v := range slice {
		if v != x {
			out = append(out, v)
		}
	}
	return out
}

// autoPromoteOneExcluding picks one member from the list, excluding excludeDID (e.g. the one who just left/demoted). Never leave 0 owners.
func autoPromoteOneExcluding(members []string, excludeDID string) []string {
	for _, m := range members {
		if m != excludeDID {
			return []string{m}
		}
	}
	return nil
}

File: ts/linkself/src/group.ts
/**
 * Group concept: member set (2+), owners, leave, dissolve, auto-promote.
 * Port of core/internal/group (Go). Storage is interface-based; domain
 * rules live here.
 */

/** Group identity, members, and owners. */
export interface Group {
  /** Unique group id. */
  id: string;
  /** Member DIDs, at least 2 while the group exists. */
  members: string[];
  /** Owner DIDs, subset of members. */
  owners: string[];
}

export type GroupErrorCode =
  | "group_not_found"
  | "too_few_members"
  | "not_member"
  | "not_owner"
  | "cannot_demote_other_owner"
  | "target_not_member"
  | "no_owner"
  | "already_member";

const GROUP_ERROR_MESSAGES: Record<GroupErrorCode, string> = {
  group_not_found: "group not found",
  too_few_members: "group requires at least 2 members",
  not_member: "not a member of the group",
  not_owner: "not an owner of the group",
  cannot_demote_other_owner: "owner cannot demote another owner",
  target_not_member: "target is not a member",
  no_owner: "cannot add member when group has no owners",
  already_member: "already a member of the group",
};

export class GroupError extends Error {
  constructor(readonly code: GroupErrorCode) {
    super(GROUP_ERROR_MESSAGES[code]);
    this.name = "GroupError";
  }
}

/** Interface for persisting group data. */
export interface GroupStore {
  /** Group IDs that include the given member DID. */
  listGroupIDsForMember(memberDID: string): Promise<string[]>;
  /** The group by ID, or null if not found. */
  getGroup(groupId: string): Promise<Group | null>;
  /** Save a new group; the implementation generates the ID. */
  createGroup(members: string[], ownerDIDs: string[]): Promise<string>;
  /** Update members and owners for the group. */
  updateGroup(groupId: string, members: string[], ownerDIDs: string[]): Promise<void>;
  /** Remove the group (e.g. when dissolved). */
  deleteGroup(groupId: string): Promise<void>;
}

/** Applies domain rules (member count, leave, owner permissions, auto-promote). */
export class GroupService {
  constructor(private readonly store: GroupStore) {}

  /** Create a group with at least 2 members. Owners can be 0, 1, or all. */
  async createGroup(members: string[], ownerDIDs: string[]): Promise<string> {
    if (members.length < 2) {
      throw new GroupError("too_few_members");
    }
    return await this.store.createGroup(members, ownerDIDs);
  }

  /** Remove the member. If the group would drop below 2 members, it is dissolved. */
  async leave(groupId: string, memberDID: string): Promise<void> {
    const g = await this.mustGet(groupId);
    if (!g.members.includes(memberDID)) {
      throw new GroupError("not_member");
    }
    const newMembers = g.members.filter((m) => m !== memberDID);
    let newOwners = g.owners.filter((o) => o !== memberDID);
    if (newMembers.length < 2) {
      await this.store.deleteGroup(groupId);
      return;
    }
    if (newOwners.length === 0) {
      newOwners = autoPromoteOneExcluding(newMembers, "");
    }
    await this.store.updateGroup(groupId, newMembers, newOwners);
  }

  /** Remove targetDID from the group. Caller must be an owner. */
  async kick(groupId: string, actorDID: string, targetDID: string): Promise<void> {
    const g = await this.mustGet(groupId);
    if (!g.owners.includes(actorDID)) {
      throw new GroupError("not_owner");
    }
    if (!g.members.includes(targetDID)) {
      throw new GroupError("target_not_member");
    }
    const newMembers = g.members.filter((m) => m !== targetDID);
    let newOwners = g.owners.filter((o) => o !== targetDID);
    if (newMembers.length < 2) {
      await this.store.deleteGroup(groupId);
      return;
    }
    if (newOwners.length === 0) {
      newOwners = autoPromoteOneExcluding(newMembers, "");
    }
    await this.store.updateGroup(groupId, newMembers, newOwners);
  }

  /** Add targetDID to owners. Caller must be an owner; target must be a member. */
  async appointOwner(groupId: string, actorDID: string, targetDID: string): Promise<void> {
    const g = await this.mustGet(groupId);
    if (!g.owners.includes(actorDID)) {
      throw new GroupError("not_owner");
    }
    if (!g.members.includes(targetDID)) {
      throw new GroupError("target_not_member");
    }
    if (g.owners.includes(targetDID)) {
      return;
    }
    await this.store.updateGroup(groupId, g.members, [...g.owners, targetDID]);
  }

  /**
   * Remove actorDID from owners. If that was the last owner, one remaining
   * member is auto-promoted.
   */
  async selfDemote(groupId: string, actorDID: string): Promise<void> {
    const g = await this.mustGet(groupId);
    if (!g.owners.includes(actorDID)) {
      throw new GroupError("not_owner");
    }
    let newOwners = g.owners.filter((o) => o !== actorDID);
    if (newOwners.length === 0) {
      newOwners = autoPromoteOneExcluding(g.members, actorDID);
    }
    await this.store.updateGroup(groupId, g.members, newOwners);
  }

  /**
   * Remove targetDID from owners. Caller must be an owner. Demoting another
   * owner is not allowed; demoting oneself delegates to selfDemote.
   */
  async demoteOwner(groupId: string, actorDID: string, targetDID: string): Promise<void> {
    const g = await this.mustGet(groupId);
    if (!g.owners.includes(actorDID)) {
      throw new GroupError("not_owner");
    }
    if (actorDID === targetDID) {
      await this.selfDemote(groupId, actorDID);
      return;
    }
    if (g.owners.includes(targetDID)) {
      throw new GroupError("cannot_demote_other_owner");
    }
  }

  /** Add newMemberDID to the group. Fails if the group has no owners. */
  async addMember(groupId: string, newMemberDID: string): Promise<void> {
    const g = await this.mustGet(groupId);
    if (g.owners.length === 0) {
      throw new GroupError("no_owner");
    }
    if (g.members.includes(newMemberDID)) {
      throw new GroupError("already_member");
    }
    await this.store.updateGroup(groupId, [...g.members, newMemberDID], g.owners);
  }

  private async mustGet(groupId: string): Promise<Group> {
    const g = await this.store.getGroup(groupId);
    if (g == null) {
      throw new GroupError("group_not_found");
    }
    return g;
  }
}

/** Pick one member (excluding excludeDID) so a group is never left with 0 owners. */
function autoPromoteOneExcluding(members: string[], excludeDID: string): string[] {
  for (const m of members) {
    if (m !== excludeDID) {
      return [m];
    }
  }
  return [];
}

/** In-memory GroupStore for tests and validation. */
export class MemGroupStore implements GroupStore {
  private readonly groups = new Map<string, Group>();

  async listGroupIDsForMember(memberDID: string): Promise<string[]> {
    const out: string[] = [];
    for (const g of this.groups.values()) {
      if (g.members.includes(memberDID)) {
        out.push(g.id);
      }
    }
    return out;
  }

  async getGroup(groupId: string): Promise<Group | null> {
    const g = this.groups.get(groupId);
    return g == null ? null : { id: g.id, members: [...g.members], owners: [...g.owners] };
  }

  async createGroup(members: string[], ownerDIDs: string[]): Promise<string> {
    const id = crypto.randomUUID();
    this.groups.set(id, { id, members: [...members], owners: [...ownerDIDs] });
    return id;
  }

  async updateGroup(groupId: string, members: string[], ownerDIDs: string[]): Promise<void> {
    const g = this.groups.get(groupId);
    if (g == null) {
      throw new GroupError("group_not_found");
    }
    this.groups.set(groupId, { id: groupId, members: [...members], owners: [...ownerDIDs] });
  }

  async deleteGroup(groupId: string): Promise<void> {
    this.groups.delete(groupId);
  }
}
